tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java

```diff
@@ -0,0 +1,304 @@
+    return removeSqueezableDimensions(tf, labels, predictions, 0);
+  }
+
+  /**
+   * Squeeze last dim if ranks differ from expected by exactly 1.
+   *
+   * @param tf the TensorFlowOps
+   * @param labels Label values, a `Tensor` whose dimensions match `predictions`.
+   * @param predictions Predicted values, a `Tensor` of arbitrary dimensions.
+   * @param expectedRankDiff Expected result of `rank(predictions) - rank(labels)`.
+   * @return `labels` and `predictions`, possibly with last dim squeezed.
+   */
+  public static <T extends TNumber> Tuple<T> removeSqueezableDimensions(
+      Ops tf, Operand<T> labels, Operand<T> predictions, int expectedRankDiff) {
+
+    tf = tf.withSubScope("removeSqueezableDimensions");
+    Shape predictionsShape = predictions.asOutput().shape();
+    int predictionsRank = predictionsShape.numDimensions();
+    Shape labelsShape = labels.asOutput().shape();
+    int labelsRank = labelsShape.numDimensions();
+
+    if (predictionsRank != Shape.UNKNOWN_SIZE && labelsRank != Shape.UNKNOWN_SIZE) {
+      // Use static rank.
+      int rankDiff = predictionsRank - labelsRank;
+      if (rankDiff == expectedRankDiff + 1 && Shape.isCompatible(predictionsShape.size(-1), 1)) {
+        predictions = tf.squeeze(predictions);
+      } else if (rankDiff == expectedRankDiff - 1 && Shape.isCompatible(labelsShape.size(-1), 1)) {
+        labels = tf.squeeze(labels);
+      }
+      return new Tuple<>(labels, predictions);
+    }
+    // Use dynamic rank.
+
+    // TODO Operand<TInt32> rankDiff = tf.math.sub(tf.rank(predictions), tf.rank(labels));
+    if (predictionsRank == Shape.UNKNOWN_SIZE && Shape.isCompatible(predictionsShape.size(-1), 1)) {
+      /*
+       * TODO, if we ever get a select that does lazy evaluation, but for now do the tf.squeeze
+       * predictions = tf.select( tf.math.equal(tf.constant(expectedRankDiff+1),rankDiff ),
+       * tf.squeeze(predictions, Squeeze.axis(Arrays.asList(-1L))), predictions ); *
+       */
+      predictions = tf.squeeze(predictions, Squeeze.axis(Collections.singletonList(-1L)));
+    }
+    if (labelsRank == Shape.UNKNOWN_SIZE && Shape.isCompatible(labelsShape.size(-1), 1)) {
+      /*
+       * TODO, if we ever get a select that does lazy evaluation labels = tf.select(
+       * tf.math.equal(tf.constant(expectedRankDiff+1),rankDiff ), tf.squeeze(labels,
+       * Squeeze.axis(Arrays.asList(-1L))), predictions ); *
+       */
+      labels = tf.squeeze(labels, Squeeze.axis(Collections.singletonList(-1L)));
+    }
+    return new Tuple<>(labels, predictions);
+  }
+
+  /**
+   * Computes the weighted loss
+   *
+   * @param tf the TensorFlow Ops
+   * @param loss the unweighted loss
+   * @param reduction the type of reduction
+   * @param sampleWeight the sample weight, if null then this defaults to one.
+   * @param <T> the data type of the loss
+   * @return the weighted loss
+   */
+  public static <T extends TNumber> Operand<T> computeWeightedLoss(
```

Comment:
This is another case where the combination of `squeezeOrExpandDimensions` and broadcasting yields a complex relationship between shapes of `sampleWeight`, `loss`, and the return value. In particular, due to broadcasting in `mul`, in the case of `reduction` == `NONE` and a surprising shape of `sampleWeight`, the return value may have a very different shape than `loss`.